router/core/src/main/scala/com/twitter/finagle/naming/buoyant/RichConnectionFailedException.scala

```diff
@@ -0,0 +1,122 @@
+package com.twitter.finagle.naming.buoyant
+
+import com.twitter.finagle.buoyant.Dst
+import com.twitter.finagle.naming.NameInterpreter
+import com.twitter.finagle._
+import com.twitter.util.Future
+import io.buoyant.namer.{DelegateTree, Delegator}
+import java.net.SocketAddress
+import scala.util.control.NoStackTrace
+
+/**
+ * An exception indicating that a connection could not be established to a remote peer.  This
+ * exception also includes additional information about the routing context for the purposes of
+ * displaying a more useful error message.  Since this exception is typically thrown in a router's
+ * client stack, the service name is not available and the `withPath` method can be used to
+ * argument this exception with the service name.
+ */
+case class RichConnectionFailedException(
+  client: Option[Name.Bound],
+  remote: SocketAddress,
+  namer: NameInterpreter
+) extends Exception(null, null)
+  with NoStackTrace
+  with SourcedException {
+
+  def withPath(service: Dst.Path): Future[Nothing] = {
+    RichConnectionFailedExceptionWithPath(
+      service,
+      client,
+      remote,
+      namer
+    )
+  }
+}
+
+/**
+ * RichConnectionFailedExceptionWithPath is a RichConnectionFailedException augmented with the
+ * service name and can be formatted as a user facing error message.
+ */
+class RichConnectionFailedExceptionWithPath(
+  service: Dst.Path,
+  client: String,
+  addresses: Set[Address],
+  remote: String,
+  resolution: Seq[String]
+) extends Exception(null, null)
+  with NoStackTrace
+  with SourcedException {
+
+  private[this] val addressList = addresses.collect {
+    case Address.Inet(isa, _) => isa.toString.stripPrefix("/")
+  }.mkString("[", ", ", "]")
+
+  private[this] val resolutionList = resolution.mkString("\n")
+
+  override def exceptionMessage(): String =
+    s"""Unable to establish connection to $remote.
+
+service name: ${service.path.show}
+client name: $client
+addresses: $addressList
+selected address: ${remote.toString.stripPrefix("/")}
+dtab resolution:
+$resolutionList
+"""
+}
+
+object RichConnectionFailedExceptionWithPath {
+
+  def apply(
+    path: Dst.Path,
+    client: Option[Name.Bound],
+    remote: SocketAddress,
+    namer: NameInterpreter
+  ): Future[Nothing] = {
+
+    val addresses = client match {
```

Comment:
This code is very similar to [formatCtx](https://github.com/linkerd/linkerd/pull/1988/files#diff-7f2b008c320604fb7424f5702567777cR16). Might be a good idea to DRY up some of the code like `val resolution` or `val addresses`. I don't see this as a PR blocker.

Reply:
possibly, although this code is short enough that I'm not sure how much there is to be gained by factoring it out.